jsonargparse_tests/test_dataclass_like.py

```diff
@@ -468,9 +474,111 @@ def test_add_class_final(parser):
     pytest.raises(ValueError, lambda: parser.add_class_arguments(FinalClass, "a", default=FinalClass()))
 
 
+if type_alias_type:
+    IntOrString = type_alias_type("IntOrString", Union[int, str])
+
+    @dataclasses.dataclass
+    class DataClassWithAliasType:
+        p1: IntOrString  # type: ignore[valid-type]
+
+    def test_bare_alias_type(parser):
+        parser.add_argument("--data", type=IntOrString)
+        cfg = parser.parse_args(["--data=MyString"])
+        assert cfg.data == "MyString"
+        cfg = parser.parse_args(["--data=3"])
+        assert cfg.data == 3
+
+    def test_dataclass_with_alias_type(parser):
+        parser.add_argument("--data", type=DataClassWithAliasType)
+        cfg = parser.parse_args(["--data.p1=MyString"])
+        assert cfg.data.p1 == "MyString"
+        cfg = parser.parse_args(["--data.p1=3"])
+        assert cfg.data.p1 == 3
+
+    @pytest.mark.skipif(not annotated, reason="Annotated is required")
+    def test_annotated_alias_type(parser):
+        parser.add_argument("--data", type=annotated[IntOrString, 1])
+        cfg = parser.parse_args(["--data=MyString"])
+        assert cfg.data == "MyString"
+        cfg = parser.parse_args(["--data=3"])
+        assert cfg.data == 3
+
+    if annotated:
+
+        @dataclasses.dataclass
+        class DataClassWithAnnotatedAliasType:
+            p1: annotated[IntOrString, 1]  # type: ignore[valid-type]
+
+    @pytest.mark.skipif(not annotated, reason="Annotated is required")
+    def test_dataclass_with_annotated_alias_type(parser):
+        parser.add_argument("--data", type=DataClassWithAnnotatedAliasType)
+        cfg = parser.parse_args(["--data.p1=MyString"])
+        assert cfg.data.p1 == "MyString"
+        cfg = parser.parse_args(["--data.p1=3"])
+        assert cfg.data.p1 == 3
+
+
 # pydantic tests
+if annotated and pydantic_support > 1:
+    import pydantic
+
+    @pydantic.dataclasses.dataclass(frozen=True)
+    class InnerDataClass:
+        """DataClass description
+
+        Args:
+            a1: a1 help
+        """
+
+        a1: int = 1
```

Comment:
Better to no use the same name in a nested and parent parameters, `a1` here and `a1` in the classes below. If they are different then it it is clearer what comes from where. For example it could be here `i1` (for inner).